src/Microsoft.Diagnostics.NETCore.Client/HandleableCollection.cs

```diff
@@ -0,0 +1,316 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Diagnostics.NETCore.Client
+{
+    /// <summary>
+    /// A collection of objects that allows for observability and mutability using handlers.
+    /// </summary>
+    internal class HandleableCollection<T> : IEnumerable<T>, IDisposable
+    {
+        public delegate bool Handler(in T item, out bool removeItem);
+
+        public delegate bool Predicate(in T item);
+
+        /// <summary>
+        /// Accepts the first item it encounters and requests that the item is removed from the collection.
+        /// </summary>
+        private static readonly Handler DefaultHandler = (in T item, out bool removeItem) => { removeItem = true; return true; };
+
+        private readonly CancellationTokenSource _disposalSource = new CancellationTokenSource();
+        private readonly List<T> _items = new List<T>();
+        private readonly List<Tuple<TaskCompletionSource<T>, Handler>> _handlers = new List<Tuple<TaskCompletionSource<T>, Handler>>();
+
+        private bool _disposed = false;
+
+        /// <summary>
+        /// Returns an enumerator that iterates through the underlying collection.
+        /// </summary>
+        /// <remarks>
+        /// The returned enumerator is over a copy of the underlying collection so that there are no concurrency issues.
+        /// </remarks>
+        IEnumerator IEnumerable.GetEnumerator()
+        {
+            VerifyNotDisposed();
+
+            IList<T> copy;
+            lock (_items)
+            {
+                copy = _items.ToList();
+            }
+            return copy.GetEnumerator();
+        }
+
+        /// <summary>
+        /// Returns an enumerator that iterates through the underlying collection.
+        /// </summary>
+        /// <remarks>
+        /// The returned enumerator is over a copy of the underlying collection so that there are no concurrency issues.
+        /// </remarks>
+        IEnumerator<T> IEnumerable<T>.GetEnumerator()
+        {
+            VerifyNotDisposed();
+
+            IList<T> copy;
+            lock (_items)
+            {
+                copy = _items.ToList();
+            }
+            return copy.GetEnumerator();
+        }
+
+        /// <summary>
+        /// Disposes the <see cref="HandleableCollection{T}"/> by clearing all items and handlers.
+        /// </summary>
+        /// <remarks>
+        /// All pending handlers with throw <see cref="ObjectDisposedException"/>.
+        /// </remarks>
+        public void Dispose()
+        {
+            if (!_disposed)
+            {
+                _disposalSource.Cancel();
```

Comment:
You need to check disposed under a lock.

Reply:
_disposalSource has now been removed.